### ISSUE TITLE
SQLite errors when working with Swift Tasks to access/modify database.

### ISSUE THREAD

**Issue by lutzmi** [ZeeZide/Enlighter#1]
Hello!

While using Lighter/Code for SQLite3/SQLite, stumbled upon some situations where SQLite itself would emit errors (disk i/o and database is locked).

I was able to reproduce the issues with a test case (see below or attached Xcode project). I would assume that all test cases should work, but only the last one, where everything is isolated via `transaction(mode: .exclusive)` runs without issues.

Not sure if I am "holding it wrong" or if it's a real bug, so any insights on this are highly appreciated :-).

Cheers, Michael

[SQLiteIOError.zip](https://github.com/ZeeZide/Enlighter/files/9777066/SQLiteIOError.zip)

**Comment by helje5** [comment]
Yes, you should never spin up 500k concurrent I/O operations, SQLite or whatever. That is extremely inefficient and inappropriate to do. Do you have a real world use case for this?

The key reason you are running into this is that I/O operations are not supposed to run on the cooperative pool. Which is why Lighter fans them out to a DispatchQueue: https://github.com/Lighter-swift/Lighter/blob/5e90f3ae08e22d740102bf09659ca670c66aa1f3/Sources/Lighter/Operations/SQLDatabaseAsyncOperations.swift#L27

That defaults to the `.global` queue, which will happily create tons of threads for you, which are going to race against each other for the SQLite database lock, which will eventually time out :-)
This works OK for a reasonable size of concurrency, but as with any `DispatchQueue.global`, you'll quickly overload the system when used inappropriately.

You can limit the actual I/O concurrency in Lighter to 1 by using a serial queue in your database, e.g  like so:
```swift
public struct TestDB : SQLDatabase, SQLDatabaseAsyncChangeOperations, SQLCreationStatementsHolder {
  public let asyncDatabaseQueue = DispatchQueue(label: "db.queue")
```
This would spool up your 500k requests and execute them in order (i.e. you won't run into I/O locking issues). It is something other SQLite libs often do by default, but you loose perfectly valid concurrency w/ that (when using SQLite in WAL mode).

But again, that would still not be a good idea (you don't want 500k pending ops hanging around either).

For write operations like your insert, serialise it in a transaction. SQLite can only do one writer at a time anyways! Doing things in a TX is much faster for a set of reasons:
```swift
    let amount = 5000
    try await db.transaction { tx in
      for i in 1 ... amount {
        let person = TestDB.Person(firstName: "FirstName\(i)", lastName: "LastName\(i)", age: 42)
        try tx.insert(person)
      }
    }
```
Same for update.

For concurrent reads the situation is a little more "it depends". In non-WAL mode you can just use the serial dispatch queue above, as it is single-reader/single-writer.
In WAL mode, you can run concurrent reads (and 1-writes) perfectly fine, but you still shouldn't saturate the system.

It would be easier to give a good recommendation for an actual use case. Neither Lighter nor async/await release one from I/O constraints :-)

P.S. Many people prefer to not use async/await for I/O but rather use specific dispatch queues (or operation queues) to run the actual I/O, to have better control over I/O concurrency.

**Comment by lutzmi** [comment]
Thanks for the quick response!

Let me maybe describe a little bit the real world scenario first, since it does not involve 500k concurrent operations (I would actually assume it’s less than 10):
To update the UI if something changes on the DB, the UI can “subscribe” to an AsyncStream for changes for either one record or a table. So in the case where the problem occurs:
- An entity gets deleted
- Which directly triggers some DB fetches from the subscribed UI
- And also additional writes are happening directly after the delete, to update a relationship table. 
This is for sure not the right way to handle relationship updates, but I would think that what is happening there is still “reasonable concurrency”.

Seeing your remark regarding WAL mode I am wondering, if switching on WAL mode would help in my case? (Not the testcase scanrio with the unreasonable amount of concurrency ;-). 

Thank you for your help!

**Comment by helje5** [comment]
WAL will definitely help w/ read concurrency (as multiple readers can acquire a read lock w/o having to wait and eventually time out).
But maybe just try to use the serial dispatch queue first.

Maybe you can group the relationship updates in a transaction?

Generally I'd avoid doing I/O based on fine grained events. Ideally one change event should result in exactly one transaction. I.e. you might want to collect the changes and apply them together at the end (which is what a CoreData managed object context does as well).

**Comment by helje5** [comment]
Wrt to reasonable concurrency: I'd say that for WAL it is fine to have one background writer synching things and one concurrent reader serving the frontend. Even sth like 10 concurrent readers _at the database level_ is _usually_ bad practice (I can imagine a few setups where this is OK, but they are few). Each concurrent access needs to open a DB connection, which is far from free and not a good idea, even w/ WAL.

Don't trigger I/O in uncoordinated ways, it is an expensive resource.

You may have convinced me to make the default async queue a single threaded one :-)

**Comment by lutzmi** [comment]
Yeah, I was under the false impression, that using the async/await APIs was safe, at least in the sense no error would occur. So I think defaulting to single threaded queue, might be safer for "novice" users like me ;-). And those who know what they're doing can change it anyhow.

As I am experimenting with Lighter/SQLite to move away from Core Data (which I was using for the last 10 years or so - but it seems it does not get the love it needs from Apple anymore, so I am looking for alternatives), I want to try to be as lightweight as possible and so I am shying away from something like a context etc. For now at least... :-).

So for now using the serial queue and coalescing more calls into transactions (which is also better for undo support I guess), seems to work fine. Thanks again for your help!
